### PR TITLE
nixos/guix: init

### PR DESCRIPTION
**Pull request by foo-dogsquared**
## Description of changes

Now that #246975 is merged, at least why not make a service module for it. While some parts of the integration test is rendered redundant once the package enables checking (e.g., `doCheck = true`), I guess it's better than nothing.

cc @cafkafk 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by foo-dogsquared** [comment]
There are still some setups I have to figure out so it's in draft for now.

File: nixos/modules/services/misc/guix/default.nix
Comment (cafkafk) [inline nixos/modules/services/misc/guix/default.nix:116]:
```suggestion
        Changing it to something other than the default will rebuild the package.
```

**Review by cafkafk — CHANGES_REQUESTED** [review]
Ohh nice work, can't wait to try this out! 

Maybe since you've done most of the work on guix you'd want to add yourself to the maintainers on the `guix` package as well?

**Comment by h7x4** [comment]
Note that there is also #264323. Maybe we could try to combine the efforts somehow?

File: nixos/modules/services/misc/guix/default.nix
Comment (h7x4) [inline nixos/modules/services/misc/guix/default.nix:151]: Examples without references that might evaluate unintentionally (like `pkgs.*`, `lib.*`, `config.*`, `options.*`) can just stay nonliteral.

```suggestion
        example = [ "--compression=zstd:6" "--repl" "--discover=no" ]
```

File: nixos/modules/services/misc/guix/default.nix
Comment (h7x4) [inline nixos/modules/services/misc/guix/default.nix:185]:
```suggestion
      users.groups.${cfg.group} = { };
```

File: nixos/modules/services/misc/guix/default.nix
Comment (h7x4) [inline nixos/modules/services/misc/guix/default.nix:91]:
```suggestion
      type = types.path;
```

File: nixos/modules/services/misc/guix/default.nix
Comment (h7x4) [inline nixos/modules/services/misc/guix/default.nix:67]:
```suggestion
      type = types.ints.unsigned;
```

File: nixos/modules/services/misc/guix/default.nix
Comment (h7x4) [inline nixos/modules/services/misc/guix/default.nix:64]: I don't see any value in making this configurable.

**Comment by foo-dogsquared** [comment]
> Maybe since you've done most of the work on guix you'd want to add yourself to the maintainers on the `guix` package as well?

Sure. That's fine by me.

> Note that there is also #264323. Maybe we could try to combine the efforts somehow?

I'm open for ideas. (I find it amusing both PRs are drafted just within an hour)

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:64]: Now that I think about it, yeah. `guixbuilder` is pretty specific anyways.

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:91]: For clarity, this also applies to `stateDir` option, right?

Comment (h7x4) [reply nixos/modules/services/misc/guix/default.nix:91]: Yeah, seems so.

I was made aware recently, that some paths used in systemd services can start with variables like `%d`, which gets invalidated by `types.path`. But `stateDir` doesn't seem to be used in this way here, so `types.path` is probably fine.

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:116]: My grammar is not the greatest so good catch!

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:151]: Got it.

**Comment by foo-dogsquared** [comment]
> Well my draft intended to open the discussion for the options, but it seems you already have the module ready.

I only upstreamed the module from [my Guix overlay](https://github.com/foo-dogsquared/nix-overlay-guix) and haven't updated or gave the module options a second look before this PR so there's still room for discussion.

**Comment by foo-dogsquared** [comment]
A few test setups later and it is done. Should be ready for review.

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:91]:
```suggestion
        This will also recompile all packages because the normal cache no longer applies.
```

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:118]:
```suggestion
          Whether to generate signing keys in {file}`/etc/guix` which
```

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:119]:
```suggestion
          are required to initialize a substitute server. Otherwise
```

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:120]:
```suggestion
          `--public-key=$FILE` and `--private-key=$FILE` can be passed in
```

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:132]:
```suggestion
          Port of the substitute server to listen on.
```

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:138]: The point of nobody is, that nobody is using it. This must be changed.

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:168]: Which? The ones from example?

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:193]:
```suggestion
```
This is self explanatory

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:201]: Why is this not dependend on some option?

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:271]: This is missing the why

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:293]: Then don't use DynamicUser? Or is anyway likely not a good choice for a daemon that stores lots of data.

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:313]: Those 3 lines can be combined into: generate keys if they are missing

File: nixos/modules/services/misc/guix/default.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/guix/default.nix:332]: Isn't this the default?

**Review by SuperSandro2000 — COMMENTED** [review]
Please read through the descriptions and make sure they are written in a style suitable for a manual.
Also please check that the comments explain the reason something is done and don't just  elaborately describe the line below with little extra information.

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:168]: Yes. It could be made clearer by linking it to [the manual page](https://guix.gnu.org/en/manual/en/html_node/Invoking-guix-gc.html).

```suggestion
          often inconvenient so it is recommended to set [some options](https://guix.gnu.org/en/manual/en/html_node/Invoking-guix-gc.html).
```

What do you think?

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:201]: Guix uses Avahi (through guile-avahi) for the autodiscovery and network service advertising of the substitute server.

Updated the comment above it which is pretty bad (now that I re-read it).

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:271]: Updated the comment.

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:138]: Updated to use `guix-publish` user instead.

**Comment by viperML** [comment]
- The nix-daemon.nix uses `unitConfig.RequiresMountsFor = "/nix/store";` , should we do the same for `/gnu/store` ?

- And more of a general guix usage question, how are channels handled? The package already comes with some channels bundled?

**Comment by foo-dogsquared** [comment]
> Please read through the descriptions and make sure they are written in a style suitable for a manual. Also please check that the comments explain the reason something is done and don't just elaborately describe the line below with little extra information.

Yeah, I neglected checking for the documentation on this one. My bad.

Will keep that in mind for future modules.

**Comment by foo-dogsquared** [comment]
> * The nix-daemon.nix uses `unitConfig.RequiresMountsFor = "/nix/store";` , should we do the same for `/gnu/store` ?

Good point. Will update the unit with those.

>  * And more of a general guix usage question, how are channels handled? The package already comes with some channels bundled?

If you meant managing custom Guix channels, this module does not handle Guix channels. You should be able to handle them as [described in the manual](https://guix.gnu.org/en/manual/en/html_node/Channels.html). As for the package, it doesn't bundle any channels.

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:332]: Yeah, just realized that.

File: nixos/modules/services/misc/guix/default.nix
Comment (foo-dogsquared) [reply nixos/modules/services/misc/guix/default.nix:293]: Yeah, I decided not to use DynamicUser= for this one and just make it run with root and with minimal hardening instead.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (wegank) [inline nixos/doc/manual/release-notes/rl-2311.section.md:67]: This should now be moved to `nixos/doc/manual/release-notes/rl-2405.section.md`.

**Review by wegank — APPROVED** [review]

**Review by cafkafk — APPROVED** [review]

**Review by cafkafk — APPROVED** [review]

**Comment by foo-dogsquared** [comment]
This has been backported to 23.11 (#272632). Now, it's a matter of waiting for it to land in `nixos-23.11` branch (or whatever stable branch you're using).

----

Also I didn't know you can backport module additions. Nice.

**Comment by jian-lin** [comment]
> Meanwhile, this makes the change in the 24.05 release note obsolete.

Could you elaborate the meaning of this and what action should be taken for https://github.com/NixOS/nixpkgs/blob/55ca7939944136731b34316a4c6fc2d4a11da122/nixos/doc/manual/release-notes/rl-2405.section.md?plain=1#L17.

**Comment by wegank** [comment]
I'd like to drop this line and, if possible, move it to the 23.11 release note.

**Pull request merged**
